API cleanup 2: make state-returning methods return references consistently

sherm1 opened the pull request:
This is a followup to #7312. Here the various state-returning APIs are switched to return references consistently. Also, the Simulator now returns Context references consistently.

As before the System convention followed here is to prefer reference return values over pointer returns unless a nullptr return is meaningful. This eliminates unnecessary nullptr checks and makes better use of the style guide's `&output_param` convention. Changes are minor but ubiquitous.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7337)
<!-- Reviewable:end -->

sherm1 commented:
Feature and platform: @soonho-tri per rotation (sorry!). Please let me know if you think this needs another feature reviewer. The only substantive changes here are to homogenize the return value types of several System infrastructure methods so that *references* to state and underlying vectors are consistently used in favor of *pointers*. Everything else is just punctuation fallout. See API changes in systems/framework/[continuous_]state.h, context.h, {leaf,diagram}_context.h, and systems/analysis/simulator.h.

There is no intent here to homogenize all of Drake, just the System infrastructure API to fix inconsistencies noted by Russ. Better to rip off the band aid fast and get this over with, I think!  

---

Review status: 0 of 195 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-:-KxOJE9ICHvd9vM4gm8D:ben14nr)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
+@soonho-tri (to put it on my reviewable plate)

---

Review status: 0 of 195 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-:-KxOe_bX5OYt4N3G5pUm:bo8b0p9)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@soonho-tri, this commit has strangely incorporated some of @mitiguy's (affecting `multibody/benchmarks/*kuka*`). In fact git is attributing the commit to "@mitiguy with @sherm1". It almost looks like a commit SHA conflict though I know that's impossible! Anyway, those changes should not be part of this PR. Please ignore those changes -- I'll figure out how to untangle this in the morning.

soonho-tri commented:
I see. Please let me know when it's ready for a review. Thank you.

---

Review status: 0 of 195 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-:-KxTBEs6TWd--6_be-Ba:b-xbxad8)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@soonho-tri, spurious changes removed -- good news is PR is half the size now in lines modified. Bad news is there are still lots of files touched. Ready for review now.

soonho-tri commented:
Checkpoint.

---

Reviewed 21 of 184 files at r1.
Review status: 21 of 184 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput_railcar.cc, line 393 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-KxUHf2YgYj8q4Nk-pAr:-KxUHf2YgYj8q4Nk-pAs:bjwyki1) ([raw file](https://github.com/robotlocomotion/drake/blob/8eee19a6d48f859beab2bb8c428034ae18b90823/drake/automotive/maliput_railcar.cc#L393)):*
> ```cpp
>   ContinuousState<T>& cs = next_state->get_mutable_continuous_state();
>   VectorBase<T>& cv = cs.get_mutable_vector();
>   MaliputRailcarState<T>* const next_railcar_state =
> ```

BTW, we still have this pattern of dynamic-casting a reference to a pointer (not a reference).
I'm wondering if you think they are good as it is or plan to have another PR to change them (or encourage the owners of the code to do so).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-:-KxUKm1s9xjwU0YYxHFj:by96gdd)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 6 of 184 files at r1.
Review status: 27 of 184 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/acrobot/acrobot_run_lqr.cc, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-KxUOLcayqNczIphrk6l:-KxUOLcayqNczIphrk6m:b-5b3o3j) ([raw file](https://github.com/robotlocomotion/drake/blob/8eee19a6d48f859beab2bb8c428034ae18b90823/drake/examples/acrobot/acrobot_run_lqr.cc#L58)):*
> ```cpp
> 
>   // Set an initial condition near the upright fixed point.
>   AcrobotStateVector<double>* x0 = dynamic_cast<AcrobotStateVector<double>*>(
> ```

FYI, another example where we can go further along with references. For example:

```c++
  AcrobotStateVector<double>& x0 = dynamic_cast<AcrobotStateVector<double>&>(
      acrobot_context.get_mutable_continuous_state_vector());
  x0.set_theta1(M_PI + 0.1);
  x0.set_theta2(-.1);
  x0.set_theta1dot(0.0);
  x0.set_theta2dot(0.0);
``` 

Of course, a question is whether we want to go all the way or not.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7337)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 27 of 184 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput_railcar.cc, line 393 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-KxUHf2YgYj8q4Nk-pAr:-KxUTRwIq4dSCADZdAgF:b-oudncq) ([raw file](https://github.com/robotlocomotion/drake/blob/8eee19a6d48f859beab2bb8c428034ae18b90823/drake/automotive/maliput_railcar.cc#L393)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, we still have this pattern of dynamic-casting a reference to a pointer (not a reference).
I'm wondering if you think they are good as it is or plan to have another PR to change them (or encourage the owners of the code to do so).
</blockquote></details>

Good question. My thought was that it depends whether the author thinks there is a possibility of meaningful failure of the dynamic_cast. Where I spotted them, I typically modified these if the code assumed without checking that the dynamic_cast would succeed (since the reference cast would provide better error checking). In the case here I saw that the return value was validated below (though only with an assert) and assumed the author had a reason for that. Switching to a reference cast here would change the failure from the assert message (including the failed condition) to a rather uninformative "bad_cast" message. I'm happy to defer to the author on whether that matters.

---

*[drake/examples/acrobot/acrobot_run_lqr.cc, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-KxUOLcayqNczIphrk6l:-KxUUnD8IPAAWknsQhld:bn0sqeu) ([raw file](https://github.com/robotlocomotion/drake/blob/8eee19a6d48f859beab2bb8c428034ae18b90823/drake/examples/acrobot/acrobot_run_lqr.cc#L58)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, another example where we can go further along with references. For example:

```c++
  AcrobotStateVector<double>& x0 = dynamic_cast<AcrobotStateVector<double>&>(
      acrobot_context.get_mutable_continuous_state_vector());
  x0.set_theta1(M_PI + 0.1);
  x0.set_theta2(-.1);
  x0.set_theta1dot(0.0);
  x0.set_theta2dot(0.0);
``` 

Of course, a question is whether we want to go all the way or not.
</blockquote></details>

Same comment here -- this one has already got a check for failure that is (marginally) better than the bad_cast message that would be produced using the reference cast. Personally I like what you wrote better, but I would leave the choice to the author.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7337)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 20 of 184 files at r1.
Review status: 47 of 184 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput_railcar.cc, line 393 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-KxUHf2YgYj8q4Nk-pAr:-KxUdm9lMdL1HCVlgTrt:b9t98kv) ([raw file](https://github.com/robotlocomotion/drake/blob/8eee19a6d48f859beab2bb8c428034ae18b90823/drake/automotive/maliput_railcar.cc#L393)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Good question. My thought was that it depends whether the author thinks there is a possibility of meaningful failure of the dynamic_cast. Where I spotted them, I typically modified these if the code assumed without checking that the dynamic_cast would succeed (since the reference cast would provide better error checking). In the case here I saw that the return value was validated below (though only with an assert) and assumed the author had a reason for that. Switching to a reference cast here would change the failure from the assert message (including the failed condition) to a rather uninformative "bad_cast" message. I'm happy to defer to the author on whether that matters.
</blockquote></details>

I see. Thanks for the answer!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7337)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (aside a BTW)

---

Reviewed 137 of 184 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/my_spring_mass_system.h, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-KxYTFOZBbbGPnUvZPPS:-KxYTFOZBbbGPnUvZPPT:b-2mc2bu) ([raw file](https://github.com/robotlocomotion/drake/blob/efc17fc438f0e1973bfa7a3cd23a3c03ae827552/drake/systems/analysis/test/my_spring_mass_system.h#L36)):*
> ```objc
>    * a system.
>    */
>   std::unique_ptr<DiscreteValues<T>> AllocateDiscreteVariables()
> ```

BTW, could you confirm that you wanted to include this change in this PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-:-KxYVWBRbBfyCdH8AKlv:b-i8gkkc)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Thanks @soonho-tri! Response below.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/my_spring_mass_system.h, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-KxYTFOZBbbGPnUvZPPS:-KxZcmBDBmQK0S8MPGQx:b-jjl83f) ([raw file](https://github.com/robotlocomotion/drake/blob/efc17fc438f0e1973bfa7a3cd23a3c03ae827552/drake/systems/analysis/test/my_spring_mass_system.h#L36)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, could you confirm that you wanted to include this change in this PR?
</blockquote></details>

Done. Thanks for pointing this out. Yes, I did want to include it but only because it is opportunistic fix to long-dead code that I encountered while passing through this file to make API related changes. Before I deleted it I verified that the default state allocation now contains exactly this code already. It is unlikely that this would get noticed again and even more unlikely that anyone would make a separate PR for it, so I fixed it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7337#-:-KxZeRtLko6bdvJd1SvS:b3fbrdb)*
<!-- Sent from Reviewable.io -->
